TST: Fix urlopen stubbing.

QuLogic opened the pull request:
For whatever reason, this stubbing does not do anything. I think it is because the
class setup overrides the module setup, but I did not look into it too deeply.  Renaming
 `setup` to `setup_module` and `teardown` to `teardown_module` fixes that.

tylerjereddy commented on numpy/lib/tests/test__datasource.py:
```diff
@@ -26,19 +26,14 @@ def urlopen_stub(url, data=None):
-    old_urlopen = urllib_request.urlopen
-    urllib_request.urlopen = urlopen_stub
 
+@pytest.fixture(autouse=True)
```
This is a bit risky since patching stdlib functions can break pytest itself--that's why I used the context manager. Here you'll monkeypatch all tests instead of just a few targeted ones.

QuLogic replied:
It's true it's in the stdlib, but it didn't seem like pytest would have much use for `urlopen`.

An autouse fixture in a module only applies to the module, so it is no different than the (intended effect of the) previous code. Though I'm not sure the exact scope of `setup`/`teardown`, so it's possible this fixture may run more times than that used to (since I specified no scope.)

tylerjereddy reviewed (commented):
Feel free to proceed with this fix & supersede mine in the linked PR -- note that @eric-wieser would also like to see some other tests in this module cleaned up as part of the PR & there's perhaps still some decision-making to be done re: fancy fixture usage in NumPy.

QuLogic commented:
An autouse fixture is already used in [numpy/conftest.py](https://github.com/numpy/numpy/blob/master/numpy/conftest.py#L39) and a module-specific fixture in [numpy/polynomial/tests/test_classes.py](https://github.com/numpy/numpy/blob/master/numpy/polynomial/tests/test_classes.py#L31), so I did not think there would be any issue with this one.

charris commented:
What happens if you just change `setup/teardown` to `setup_module/teardown_module`? Also looks like the class level versions should be `@classmethod` `setup_class/teardown_class` or `setup_method/teardown_method`. Need to take a closer look to understand intent.

QuLogic commented:
> What happens if you just change `setup/teardown` to `setup_module/teardown_module`?

Yes, that does appear to work.

> Also looks like the class level versions should be `@classmethod` `setup_class/teardown_class` or `setup_method/teardown_method`.

Unless the `DataSource` can be reused, I don't think it should be a class method. Don't know if adding `_method` would make much difference.

charris commented:
@QuLogic If simply fixing `setup/teardown` works, I'd prefer that fix because it is simpler and more obvious.

QuLogic committed:
TST: Fix urlopen stubbing.

setup/teardown doesn't work, but setup_module/teardown_module does, at
least in pytest.

charris commented:
Thanks @QuLogic .
